src/rpc/methods/eth.rs

```diff
@@ -1338,13 +1330,84 @@ impl RpcMethod<2> for EthGetCode {
         let get_bytecode_return: GetBytecodeReturn =
             fvm_ipld_encoding::from_slice(msg_rct.return_data().as_slice())?;
         if let Some(cid) = get_bytecode_return.0 {
-            Ok(Bytes(ctx.store().get_required(&cid)?))
+            Ok(EthBytes(ctx.store().get_required(&cid)?))
         } else {
             Ok(Default::default())
         }
     }
 }
 
+pub enum EthGetStorageAt {}
+impl RpcMethod<3> for EthGetStorageAt {
+    const NAME: &'static str = "Filecoin.EthGetStorageAt";
+    const PARAM_NAMES: [&'static str; 3] = ["eth_address", "position", "block_number_or_hash"];
+    const API_VERSION: ApiVersion = ApiVersion::V1;
+    const PERMISSION: Permission = Permission::Read;
+
+    type Params = (EthAddress, EthBytes, BlockNumberOrHash);
+    type Ok = EthBytes;
+
+    async fn handle(
+        ctx: Ctx<impl Blockstore + Send + Sync + 'static>,
+        (eth_address, position, block_number_or_hash): Self::Params,
+    ) -> Result<Self::Ok, ServerError> {
+        let make_empty_result = || EthBytes(vec![0; 32]);
+
+        let ts = tipset_by_block_number_or_hash(&ctx.chain_store, block_number_or_hash)?;
+        let to_address = FilecoinAddress::try_from(&eth_address)?;
+        let Some(actor) = ctx
+            .state_manager
+            .get_actor(&to_address, *ts.parent_state())?
+        else {
+            return Ok(make_empty_result());
+        };
+
+        if !fil_actor_interface::is_evm_actor(&actor.code) {
+            return Ok(make_empty_result());
+        }
+
+        let params = RawBytes::new(GetStorageAtParams::new(position.0)?.serialize_params()?);
+        let message = Message {
+            from: FilecoinAddress::SYSTEM_ACTOR,
+            to: to_address,
+            method_num: METHOD_GET_STORAGE_AT,
+            gas_limit: BLOCK_GAS_LIMIT,
+            params,
+            ..Default::default()
+        };
+        let mut api_invoc_result = None;
```

Comment:
nit: could be done without`mut`.

```
let api_invoice_result = {
  for ts in ... {
     match {
     Ok(res) -> return Some(res)
  }
  None
}
```

Reply:
The `return` clause works for the function instead of the scope unless the scope is converted to a lambda. I was thinking about `break Some(res);` syntax just like in some other languages but it does not work in rust

Reply:
If you want to avoid `mut` here, you still could use labelled blocks, e.g.:

```
let api_invoc_result = 'block: {
    for ts in ts.chain_arc(ctx.store()) {
        match ctx.state_manager.call(&message, Some(ts)) {
            Ok(res) => {
                break 'block res;
            }
            Err(e) => tracing::warn!(%e),
        }
    }
    return Err(anyhow::anyhow!("Call failed").into());
};
```

Reply:
@elmattic Didn't know about this. This is great, thanks!